FreeRTOS/Demo/ARC_NSIM_QEMU_GCC_MW/arc/board.c

```diff
@@ -0,0 +1,345 @@
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+   --------------------------------------------- */
+#include "arc.h"
+#include "board.h"
+#include "arc_timer.h"
+#include "arc_exception.h"
+#include <stdint.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <ctype.h>
+#include <errno.h>
+
+#define MAX_SYS_COUNTER_VALUE       (0xffffffff)
+
+#ifndef BOARD_SYS_TIMER_HZ
+#define BOARD_SYS_TIMER_HZ      (1000)
+#endif
+
+typedef struct main_args {
+	int argc;
+	char *argv[];
+} MAIN_ARGS;
+
+/** Change this to pass your own arguments to main functions */
+MAIN_ARGS s_main_args = { 1, { "main" } };
+
+/** board timer interrupt reset count */
+static uint32_t cyc_hz_count = (BOARD_CPU_CLOCK / BOARD_SYS_TIMER_HZ);
+
+/** board timer counter in timer interrupt */
+static volatile uint64_t gl_sys_hz_cnt = 0;
+/** board 1ms counter */
+static volatile uint32_t gl_ms_cnt = 0;
+
+#define HZ_COUNT_CONV(precision, base)  ((precision) / (base))
+
+extern void __mw_run_tls_dtor();
+void task_end_hook( void * pxTCB );
+
+void task_end_hook( void * pxTCB )
+{
+	TaskHandle_t task2free = ( TaskHandle_t ) pxTCB;
+
+	if( task2free != NULL )
+	{
+		void * tls = pvTaskGetThreadLocalStoragePointer( task2free, 0 );
+
+		if( tls )
+		{
+			__mw_run_tls_dtor();
+			vPortFree( tls );
+			vTaskSetThreadLocalStoragePointer( task2free, 0, NULL );
+		}
+	}
+}
+
+void gnu_printf_setup(void)
+{
+#if defined(__GNU__) && !defined(_HAVE_LIBGLOSS_) && !defined(_HOSTLINK_)
+	exc_handler_install(EXC_NO_SWI, (EXC_HANDLER_T)syscall_swi);
+#endif
+}
+
+#if defined(__GNU__)
+extern char __start_heap;
+extern char __end_heap;
+
+caddr_t *SYSCALL_PREFIX(_sbrk)(size_t nbytes)
+{
+	static char* heap_ptr = NULL;
+	char* prev_heap_ptr;
+
+	if (heap_ptr == NULL) {
+		heap_ptr = &__start_heap;
+	}
+
+	/* Align the 'heap_ptr' so that memory will always be allocated at word
+	boundaries. */
+	heap_ptr = (char *) ((((unsigned long) heap_ptr) + 7) & ~7);
+	prev_heap_ptr = heap_ptr;
+
+	if ((heap_ptr + nbytes) < &__end_heap) {
+		heap_ptr += nbytes;
+		return (caddr_t *) prev_heap_ptr;
+	}
+	errno = ENOMEM;
+	return (caddr_t *) -1;
+}
```

Comment:
Do we really use dynamic memory management (so we implement _sbrk)?
If we don't use dynamic memory management I guess we can drop it?

Reply:
In `main.c: 186`, I use `sprintf`, which uses heap internally.

Reply:
> sprintf, which uses heap internally.

Hmm, are you sure about that?

Standard `sprintf` implementation doesn't use any dynamic memory allocation. It only uses the buffer passed to it as a first argument.

Reply:
One explanation I found in [stackoverflow](https://stackoverflow.com/questions/42287206/sprintf-without-memory-allocation#:~:text=sprintf%20is%20using%20heap%20internally.%20So%20your%20best%20bet%20is%20to%20write%20a%20version%20which%20doesn%27t.%20If%20you%20just%20use%20%25d%2C%20%25x%20and%20%25s%20format%20specifiers%2C%20you%20won%27t%20need%20any%20dynamic%20memory.) .Once I remove `_sbrk`, I would got an error `in function _sbrk: sbrk.c:(.text._sbrk+0x10): undefined reference to end****`

Reply:
As I can see in FreeRTOS there is sprintf implementation (which doesn't use heap). Can we reuse it instead of toolchain one?

Reply:
switched to sprintf without heap implementation used in other FreeRTOS DEMO.